xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,186 @@
+/** @file featurelist_internal.cc
+ * @brief Definition of Feature::Internal class.
+ */
+/* Copyright (C) 2012 Parth Gupta
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/featurelist.h"
+#include "featurelist_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+void
+FeatureList::Internal::set_database(const Xapian::Database & db)
+{
+    LOGCALL_VOID(API, "FeatureList::set_database", db);
+    featurelist_db = db;
+}
+
+void
+FeatureList::Internal::set_query(const Xapian::Query & query)
+{
+    LOGCALL_VOID(API, "FeatureList::set_query", query);
+    if (query.empty()) {
+	throw Xapian::InvalidArgumentError("Can't initialise with an empty query string");
```

Comment:
> It should be invalid here as the query object with empty string

Stop thinking about query strings here.  A query object doesn't hold a query string - it's a tree of query objects representing terms, operations to combine them, etc (which is sometimes but not always obtained by parsing a query string).

A key point is that `query.empty()` tests for a **subset** of query objects which won't match anything - `Query()` aka `Query::MatchNothing`, and some forms which are reduced to this during construction - e.g. `Query() & Query("a")` happens to be currently, but these reductions are mostly opportunistic, and exactly what gets reduced has evolved with time, and is likely to continue to do so.  Meanwhile it's trivially easy to construct queries which don't get reduced to `Query()` but will never match anyway - e.g. `Query("a") & ~Query("a")`.

By checking `query.empty()` here, you reject a rather arbitrary subset of queries which can't match anything, and that may change between versions which is somewhat confusing to the user.  Why is `Query() & Query("a")` invalid, but `Query("a") & ~Query("a")` isn't?  And in a future version, why does the latter suddenly become invalid?

But there are still queries which can't match anything you can't catch here, and also queries which might match something but don't with the database being searched (e.g. `Query("foo")` if term "foo" isn't present), so with the current check you're going to have to deal with this case anyway aren't you?

I think as James suggests, the correct test is not `query.empty()` but `mset.empty()`.  If there are no results, there's nothing to rerank and you should be able to short-cut everything at a higher level.